File: 07/Java/TripleExpression/CheckedSubtract.java
package TripleExpression;

public class CheckedSubtract extends BinaryOperator implements TripleExpression {
    public CheckedSubtract(TripleExpression first, TripleExpression second) {
        super(first, second);
    }

    protected void check(int a, int b) throws Exception {
        if(b > 0 && a < Integer.MIN_VALUE + b) {
            throw new OverflowException();
        }
        if(b < 0 && a > Integer.MAX_VALUE + b) {
            throw new OverflowException();
        } 
    }

    protected int apply(int a, int b) throws Exception {
        check(a, b);
        return a - b;
    }
}

File: 07/CSharp/TripleExpressionApp/TripleExpressionApp/CheckedSubtract.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace TripleExpressionApp
{
    public class CheckedSubtract : BinaryOperator
    {
        public CheckedSubtract(ITripleExpression first, ITripleExpression second)
        {
            fOp = first;
            sOp = second;
        }

        protected void Check(double a, double b)
        {
            if (b > 0 && a < double.MinValue + b)
            {
                throw new OverflowException();
            }
            if (b < 0 && a > double.MaxValue + b)
            {
                throw new OverflowException();
            }
        }

        override protected double Apply(double a, double b)
        {
            Check(a, b);
            return a - b;
        }
    }
}
